main/zookeeper_watcher.go

```diff
@@ -185,6 +183,42 @@ func main() {
 	wg.Wait()
 }
 
+func waitForConnected(eventChannel <-chan zk.Event) {
+	stateChannel := make(chan bool)
+	quitChannel := make(chan bool)
+	go determineState(eventChannel, stateChannel, quitChannel)
+	connected := false
+	for true {
+		stateChannel <- true
+		currentState := <-stateChannel
+		log.Infof("Current Connected state = %v", currentState)
+		if currentState && connected {
+			log.Infof("State is connected twice in a row, done waiting for connect")
+			quitChannel <- true
+			return
+		}
+		connected = currentState
+		time.Sleep(time.Second)
+	}
+}
+
+func determineState(eventChannel <-chan zk.Event, stateChannel, quitChannel chan bool) {
+	connected := false
+	for true {
+		select {
+		case ev := <-eventChannel:
+			log.Infof("Event = %s", ev)
+			connected = ev.State == zk.StateConnected || ev.State == zk.StateHasSession
+			log.Infof("Zookeeper connected state is %v", connected)
+		case <-stateChannel:
+			stateChannel <- connected
+		case <-quitChannel:
```

Comment:
I feel like this would be the perfect situation to use a closed channel to tell when to end your for true loop instead of a third channel. Maybe iterate through all elements in stateChannel. https://gobyexample.com/range-over-channels